tests/unit/test_charm.py

```diff
@@ -208,6 +137,82 @@ def test_update_status_down(self, _process_wheel_file, _setup_container, mock_ex
 
         self.assertEqual(harness.model.unit.status, MaintenanceStatus("Status check: DOWN"))
 
+    @mock.patch("charm.TemporalWorkerK8SOperatorCharm._process_wheel_file")
+    @mock.patch("charm._setup_container")
+    def test_vault_relation(self, _process_wheel_file, _setup_container):
+        """The charm is ready with vault relation."""
+        harness = self.harness
+
+        state = simulate_lifecycle(harness, CONFIG)
+        harness.charm.on.config_changed.emit()
+
+        module_name = json.loads(state["module_name"])
+        unpacked_file_name = json.loads(state["unpacked_file_name"])
+        command = f"python worker.py {unpacked_file_name} {module_name}"
+
+        add_vault_relation(self, harness)
+        self.harness.update_config({})
+
+        # The plan is generated after pebble is ready.
+        want_plan = {
+            "services": {
+                "temporal-worker": {
+                    "summary": "temporal worker",
+                    "command": command,
+                    "startup": "enabled",
+                    "override": "replace",
+                    "environment": {**WANT_ENV, **EXPECTED_VAULT_ENV},
+                    "on-check-failure": {"up": "ignore"},
+                }
+            },
+            "checks": {
+                "up": {
+                    "override": "replace",
+                    "level": "alive",
+                    "period": "10s",
+                    "exec": {"command": "python check_status.py"},
+                }
+            },
+        }
+
+        got_plan = harness.get_container_pebble_plan("temporal-worker").to_dict()
+        self.assertEqual(got_plan, want_plan)
+
+
+def add_vault_relation(test, harness):
```

Comment:
It might be important to also validate that the vault relation gone away does indeed remove the vault credentials.